x-pack/agent/pkg/agent/transpiler/steps.go

```diff
@@ -0,0 +1,185 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package transpiler
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/pkg/errors"
+	"gopkg.in/yaml.v2"
+)
+
+// StepList is a container that allow the same tree to be executed on multiple defined Step.
+type StepList struct {
+	Steps []Step
+}
+
+// Step is an execution step which needs to be run.
+type Step interface {
+	Execute(rootDir string) error
+}
+
+// Execute executes a list of steps.
+func (r *StepList) Execute(rootDir string) error {
+	var err error
+	for _, step := range r.Steps {
+		err = step.Execute(rootDir)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// MarshalYAML marsharl a steps list to YAML.
+func (r *StepList) MarshalYAML() (interface{}, error) {
+	doc := make([]map[string]Step, 0, len(r.Steps))
+
+	for _, step := range r.Steps {
+		var name string
+		switch step.(type) {
+		case *DeleteFileStep:
+			name = "delete_file"
+		case *MoveFileStep:
+			name = "move_file"
+
+		default:
+			return nil, fmt.Errorf("unknown rule of type %T", step)
+		}
+
+		subdoc := map[string]Step{
+			name: step,
+		}
+
+		doc = append(doc, subdoc)
+	}
+	return doc, nil
+}
+
+// UnmarshalYAML unmarshal a YAML document into a RuleList.
+func (r *StepList) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	var unpackTo []map[string]interface{}
+
+	err := unmarshal(&unpackTo)
+	if err != nil {
+		return err
+	}
+
+	// NOTE(ph): this is a bit of a hack because I want to make sure
+	// the unpack strategy stay in the struct implementation and yaml
+	// doesn't have a RawMessage similar to the JSON package, so partial unpack
+	// is not possible.
+	unpack := func(in interface{}, out interface{}) error {
+		b, err := yaml.Marshal(in)
+		if err != nil {
+			return err
+		}
+		fmt.Println(string(b))
+		return errors.Wrap(yaml.Unmarshal(b, out), "ajaj")
+	}
+
+	var steps []Step
+
+	for _, m := range unpackTo {
+		ks := keys(m)
+		if len(ks) > 1 {
+			return fmt.Errorf("unknown rule identifier, expecting one identifier and received %d", len(ks))
+		}
+
+		name := ks[0]
+		fields := m[name]
+
+		var s Step
+		switch name {
+		case "delete_file":
+			s = &DeleteFileStep{}
+		case "move_file":
+			s = &MoveFileStep{}
+		default:
+			return fmt.Errorf("unknown rule of type %s", name)
+		}
+
+		if err := unpack(fields, s); err != nil {
+			return err
+		}
+
+		steps = append(steps, s)
+	}
+	r.Steps = steps
+	return nil
+}
+
+// DeleteFileStep removes a file from disk.
+type DeleteFileStep struct {
+	Path string
+	// FailOnMissing fails if file is already missing
+	FailOnMissing bool
+}
+
+// Execute executes delete file step.
+func (r *DeleteFileStep) Execute(rootDir string) error {
+	path := filepath.Join(rootDir, filepath.FromSlash(r.Path))
+	err := os.Remove(path)
```

Comment:
We probably want to add some defensive programming here, even if we control the spec, we never know when this will not be the case. I think we should convert the `Filepath.Join()` into an absolute path and make sure we are in the "rootdir" and not somewhere else. 

I believe this is the case, we want to sandbox every command into a specific directly. I assume all our commands operate in the extracted directory.

Reply:
this is something i was thinking about but did not want to sound too paraniod

Reply:
Better to be paranoid than sorry :)